src/khepri_lock.erl

```diff
@@ -0,0 +1,654 @@
+%% '''
+%%
+%% === Exclusive locks ===
+%%
+%% Locks can be <em>exclusive</em>: only one process can acquire it at a time.
+%% A lock can be acquired by multiple processes if all processes call {@link
+%% acquire/4} with the same `LockId' and pass the same `group' option in
+%% {@link options()}. If a lock is non-exclusive, it must also be recursive.
+%% Locks are exclusive by default.
+%%
+%% ```
+%% LockId = my_lock,
+%% Options = #{group => my_group},
+%% {ok, Lock1} = khepri_lock:acquire(LockId, local, Options),
+%% %% Another process can also acquire the lock if it provides the same LockId
+%% %% and `group' term.
+%% spawn(fun() ->
+%%               {ok, Lock2} = khepri_lock:acquire(LockId, local, Options),
+%%               true = khepri_lock:release(Lock2)
+%%       end),
+%% khepri_lock:release(Lock1)
+%% '''
+
+-module(khepri_lock).
+-behavior(gen_server).
+
+-include_lib("kernel/include/logger.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+-include("src/khepri_lock.hrl").
+-include("src/khepri_cluster.hrl").
+-include("include/khepri.hrl").
+
+-define(SERVER, ?MODULE).
+
+-define(IS_SCOPE(Scope), (Scope =:= local orelse Scope =:= cluster)).
+
+-type lock_id() :: term().
+%% Any term which represents a resource that can be locked.
+
+-type group_id() :: pid() | reference() | atom().
+%% The identifier for a lock's group.
+%%
+%% This value can take one of three types. Each type creates a lock with
+%% different recursivity and exclusitivy properties:
+%%
+%% <ul>
+%% <li>{@link reference()}: exclusive and non-recursive</li>
+%% <li>{@link pid()}: exclusive and recursive</li>
+%% <li>{@link atom()}: non-exclusive and recursive</li>
+%% </ul>
+
+-type lock() :: #khepri_lock{}.
+%% A term representing an acquired lock.
+%%
+%% {@link acquire/4} returns this type if a lock can successfully be acquired.
+%% Locks should be passed to {@link release/1} to make the lock available to
+%% by acquired by others.
+
+-type scope() :: local | cluster.
+%% The scope to which a lock is considered unique.
+%%
+%% Locks may either be held locally by the current Erlang node or cluster-wide
+%% for a Khepri store. Local locks do not involve any distribution or
+%% networking. Local locks also do not interact with any Khepri store: the
+%% `StoreId' argument to {@link acquire/4} is only used for uniqueness.
+%%
+%% Cluster-wide locks must be recognized as acquired or released by a quorum in
+%% the cluster before {@link acquire/4} will consider the lock acquired or
+%% {@link release/1} will consider the lock to be released.
+
+-type options() :: #{retries => non_neg_integer() | infinity,
+                     recursive => boolean(),
+                     group => atom(),
+                     release_on_exit => boolean()}.
+%% Options for acquiring a lock with {@link acquire/4}.
+%%
+%% <ul>
+%% <li>`retries' controls the number of times {@link acquire/4} will attempt to
+%% acquire a lock before giving up.</li>
```

Comment:
I think a timeout would be a better API but with the try->sleep->retry mechanism used in `global`, there is potential for the caller to wait longer than the timeout they provide. We could tune the backoff numbers though so that it's only possible to overshoot the timeout by some predictable margin.

Another way to approach locking would be to have a waiting queue. See https://github.com/whitfin/sleeplocks/blob/master/src/sleeplocks.erl: the gen_server process doesn't reply to a caller until the lock can be granted and locks are passed to the next waiting caller when the current holder releases the lock. This approach would let us use an exact timeout but I think this approach only works if you have an `infinity` timeout, right? If the caller timed out and then was granted the lock, it would have no way to "refuse" it and then that lock would be stuck.